File: operation/activation/gelu.py
"""Gaussian Error Linear Unit activation function

Smooth approximation of ReLU used in many modern transformer architectures.
"""
from __future__ import annotations

import torch
from torch import Tensor

from caramba.operation.activation.base import ActivationOperation


class GELUOperation(ActivationOperation):
    """Gaussian Error Linear Unit activation

    Smooth, differentiable approximation of ReLU that performs well in practice,
    commonly used in transformer models like BERT and GPT.
    """
    def __init__(self) -> None:
        super().__init__()

    def forward(self, *, x: Tensor) -> Tensor:
        """Apply GELU activation

        Computes 0.5 * x * (1 + erf(x/sqrt(2))), providing smooth non-linearity
        that outperforms ReLU in many transformer architectures.
        """
        return torch.nn.functional.gelu(x)

File: pkg/backend/compute/cpu/operation/activation/gelu.go
package activation

import "github.com/theapemachine/caramba/pkg/backend/compute/state"

/*
Gelu evaluates the tanh-form Gaussian Error Linear Unit:

	0.5 * x * (1 + tanh(sqrt(2/pi) * (x + 0.044715*x^3)))
*/
type Gelu struct{}

/*
NewGelu returns an Operation that computes GELU elementwise.
*/
func NewGelu() *Gelu {
	return &Gelu{}
}

func (gelu *Gelu) Forward(stateDict *state.Dict) (*state.Dict, error) {
	if err := stateDict.RequireOperation("activation.gelu"); err != nil {
		return nil, err
	}

	geluKernel(stateDict.Out, stateDict.Inputs[0])

	return stateDict, nil
}
